src/main/java/uk/ac/ic/wlgitbridge/bridge/context/ContextStore.java

```diff
@@ -0,0 +1,132 @@
+   * - @see GenericProjectContextFactory
+   * - @see PostgresProjectContextFactory
+   */
+  public static void Initialize(ProjectContextFactory contextFactory) throws ContextStoreAlreadyInitialisedException {
+    Log.info("Initialize ContextStore");
+    if (ContextStore.instance != null) {
+        Log.error("ContextStore already initialized");
+      throw new ContextStoreAlreadyInitialisedException();
+    }
+    ContextStore.instance = new ContextStore(contextFactory);
+  }
+
+  /**
+   * Reset the ContextStore, only to be used in tests,
+   * hence the __scary__ underscores
+   */
+  public static void __Reset__() {
+    Log.info("Reset ContextStore");
+    if (ContextStore.instance != null) {
+      ContextStore.Instance().stop();
+      ContextStore.instance = null;
+    }
+  }
+
+  public static ContextStore Instance()  {
+    if (ContextStore.instance == null) {
+      throw new RuntimeException("ContextStore not initialized");
+    }
+    return ContextStore.instance;
+  }
+
+  /**
+   * Cleanup contexts, and pass on the close signal.
+   * Essential for cleanly shutting down the server
+   */
+  public synchronized void stop() {
+    Log.info("Stopping ContextStore");
+    this.clearLocalContexts();
+    if (this.projectContextFactory != null) {
+      this.projectContextFactory.close();
+    }
+  }
+
+  private void ensureThreadLocalStorage() {
+    if (this.localContexts.get() == null) {
+      HashMap<String, ProjectContext> contextMap = new HashMap<String, ProjectContext>();
+      this.localContexts.set(contextMap);
+    }
+  }
+
+  private synchronized ProjectContext getFromThreadLocalStorage(String projectName) {
+    this.ensureThreadLocalStorage();
+    HashMap<String, ProjectContext> contextMap = this.localContexts.get();
+    return contextMap.get(projectName);
+  }
+
+  private synchronized void storeInThreadLocalStorage(String projectName, ProjectContext context) {
+    this.ensureThreadLocalStorage();
+    HashMap<String, ProjectContext> contextMap = this.localContexts.get();
+    contextMap.put(projectName, context);
+  }
+
+  public ProjectContext getContextForProject(String projectName) {
+    ProjectContext storedContext = this.getFromThreadLocalStorage(projectName);
+    if (storedContext != null) {
+      if (!storedContext.getProjectName().equals(projectName)) {
+        Log.error("Context ID does not match " + projectName + ", " + storedContext.getProjectName());
+        throw new RuntimeException("Wrong project ID");
+      }
+      return storedContext;
+    }
+    if (shuttingDown) {
+      throw new RuntimeException("ContextStore is shutting down");
+    }
+    ProjectContext newContext = this.projectContextFactory.makeProjectContext(projectName);
+    this.storeInThreadLocalStorage(projectName, newContext);
+    return newContext;
+  }
+
+  /**
+   * Clean up all contexts used on this thread. Good to call at the end of a
+   * git operation.
+   */
+  public void clearLocalContexts() {
```

Comment:
This seems like a bit of a footgun --- it could clear too much or get forgotten. It would be nice if closing the context cleaned up the map. Given that the context store is a singleton, I think the project context could unregister itself? Or we could pass an instance of the store (`this`) when creating the context.

Reply:
Yeah, we could do something like that. I'll look into it :+1:

Reply:
Looking into it, I'm not sure there's a way to make this (much) more elegant, as it's the lock that gets closed at the end of a block, rather than the context, especially now that we're grabbing the lock directly, and skipping the reference to the context object. I feel that just wiping the slate clean at the end of a logical operation is acceptable, although maybe we could move the call to the _top_ of the operation, and maybe into the Handler functions above the Bridge in the call-stack? That way we would be resetting to a good state at the start, whenever an existing thread is asked to do new work?

Reply:
> move the call to the top of the operation

That sounds worth a shot, yes. That does seem like it would be harder to get wrong.

Reply:
But, on the other hand... it's not very good to leave the postgres connection hanging around, checked out of the pool, until the next request. Where is the connection getting released back to the pool at the moment?

Reply:
Spoke too soon, I have a cunning plan... that just might work...

Reply:
Behold, the power of lambda: https://github.com/overleaf/writelatex-git-bridge/pull/93/files#diff-48e673a09e3145e50b0ac765436d2ca75028ff0daeb0a4e6c4f646b2ff8b5486R321